controllers/azuremachine_controller.go

```diff
@@ -51,9 +51,33 @@ import (
 // AzureMachineReconciler reconciles a AzureMachine object
 type AzureMachineReconciler struct {
 	client.Client
-	Log              logr.Logger
-	Recorder         record.EventRecorder
-	ReconcileTimeout time.Duration
+	Log                        logr.Logger
+	Recorder                   record.EventRecorder
+	ReconcileTimeout           time.Duration
+	azureMachineServiceFactory func(machineScope *scope.MachineScope, clusterScope *scope.ClusterScope) *azureMachineService
+}
+
+// NewAzureMachineReconciler returns a new AzureMachineReconciler instance
+func NewAzureMachineReconciler(client client.Client, log logr.Logger, recorder record.EventRecorder, reconcileTimeout time.Duration,
+	opts ...func(azureMachineReconciler *AzureMachineReconciler)) *AzureMachineReconciler {
```

Comment:
I'm good with the variadic options, but I'd prefer to have have a type for an option rather than a raw func definition.

Reply:
Do you mean something like `type option func(azureMachineReconciler *AzureMachineReconciler)`?

Reply:
Yes, here's an example of what I mean.

Definition:
https://github.com/devigned/pub/blob/8124492a2e15b32f2d0a55d9b6e7a71f33ed4478/pkg/partner/client.go#L40

Usage:
https://github.com/devigned/pub/blob/8124492a2e15b32f2d0a55d9b6e7a71f33ed4478/pkg/partner/client.go#L154

Reply:
Pending the need to use variadic options, of course.